neqo-common/src/udp.rs

```diff
@@ -101,16 +101,19 @@ impl Socket {
             return Ok(None);
         }
 
-        if meta.len == buf.len() {
-            eprintln!("Might have received more than {} bytes", buf.len());
+        if meta.len == self.recv_buf.len() {
+            eprintln!(
+                "Might have received more than {} bytes",
+                self.recv_buf.len()
+            );
         }
 
         Ok(Some(Datagram::new(
             meta.addr,
             *local_address,
             meta.ecn.map(|n| IpTos::from(n as u8)).unwrap_or_default(),
             None, // TODO: get the real TTL https://github.com/quinn-rs/quinn/issues/1749
-            &buf[..meta.len],
+            &self.recv_buf[..meta.len],
         )))
```

Comment:
Note that each call to `Datagram::new` allocates the datagram's data in a new `Vec`.

https://github.com/mozilla/neqo/blob/ad027cfe2901dc98b3dae20574349f52621c5622/neqo-common/src/datagram.rs#L20-L35

Changing this is a larger undertaking I would like to tackle with https://github.com/mozilla/neqo/issues/1693.